### PR TITLE
Settings

### PR DESCRIPTION
**Pull request by tollmanz**
This PR adds a settings panel to the Purgely plugin. The settings added to the database will override values configured via constants.

### CODE REVIEW

**Comment by tollmanz** [comment]
Just to keep myself organized, I need to complete the following:
- [x] Add settings descriptions
- [x] Add better settings page copy
- [x] Reorganize the settings page to make things more readable
- [x] Implement the settings
- [x] Get tests to pass again

**Comment by simonwistow** [comment]
Let me know if I can help with the copy or anything.

**Comment by tollmanz** [comment]
> Let me know if I can help with the copy or anything.

Hey @simonwistow! I only saw this message after I pushed up descriptions. They are not very good. You probably have a better sense of good ways to describe these settings. Please take a look!

**Comment by tollmanz** [comment]
@simonwistow I took another, much better pass, at the settings copy. Can you or someone on your team review? If it looks good, I'm ready to merge this!

File: src/settings-page.php [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+			'stale_while_revalidate_ttl',
+			__( 'Stale While Revalidate TTL (in Seconds)', 'purgely' ),
+			array( $this, 'stale_while_revalidate_ttl_render' ),
+			'purgely-settings',
+			'purgely-stale_settings'
+		);
+
+		add_settings_field(
+			'enable_stale_while_error',
+			__( 'Enable Stale While Error', 'purgely' ),
```
Comment (simonwistow) [inline src/settings-page.php]: Technically it's "Stale If Error"

File: src/settings-page.php [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+
+	/**
+	 * Print the description for the settings page.
+	 *
+	 * @since 1.0.0.
+	 *
+	 * @return void
+	 */
+	public function fastly_settings_callback() {
+		esc_html_e( 'Please enter details related to your Fastly account. A Fastly API key and service ID are required for some operations (e.g., surrogate key and full cache purges). ', 'purgely' );
```
Comment (simonwistow) [inline src/settings-page.php:189]: I'd be tempted to remove the "optional" part and make them mandatory for simplicity's sake.

File: src/settings-page.php [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+	 * @since 1.0.0.
+	 *
+	 * @return void
+	 */
+	public function stale_while_revalidate_ttl_render() {
+		$options = Purgely_Settings::get_settings();
+		?>
+		<input type='text' name='purgely-settings[stale_while_revalidate_ttl]' value='<?php echo esc_attr( $options['stale_while_revalidate_ttl'] ); ?>'>
+		<p class="description">
+			<?php esc_html_e( 'This setting determines the amount of time that stale content will be served while new content is generated.', 'purgely' ); ?>
```
Comment (simonwistow) [inline src/settings-page.php:396]: Ditto

> ... regenerated in the background.

File: src/settings-page.php [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+	 */
+	public function enable_stale_while_revalidate_render() {
+		$options = Purgely_Settings::get_settings();
+		?>
+		<input type='radio' name='purgely-settings[enable_stale_while_revalidate]' <?php checked( isset( $options['enable_stale_while_revalidate'] ) && true === $options['enable_stale_while_revalidate'] ); ?> value='true'>Yes&nbsp;
+		<input type='radio' name='purgely-settings[enable_stale_while_revalidate]' <?php checked( isset( $options['enable_stale_while_revalidate'] ) && false === $options['enable_stale_while_revalidate'] ); ?> value='false'>No
+		<p class="description">
+			<?php
+			printf(
+				esc_html__( 'Turn the "stale while revalidate" behavior on or off. The stale while revalidate behavior allows stale content to be served while content is regenerated. Please see %s', 'purgely' ),
```
Comment (simonwistow) [inline src/settings-page.php]: I think it's better phrased as 

> ... regenerated in the background.

File: purgely.php [hunk 5]
```diff
@@ -129,18 +134,18 @@ public function __construct() {
 		$this::$surrogate_keys_header = new Purgely_Surrogate_Keys_Header();
 
 		// Initialize the surrogate control header.
-		$this::$surrogate_control_header = new Purgely_Surrogate_Control_Header( PURGELY_SURROGATE_CONTROL_TTL );
+		$this::$surrogate_control_header = new Purgely_Surrogate_Control_Header( Purgely_Settings::get_setting( 'surrogate_control_ttl' ) );
 
 		// Add the surrogate keys.
 		add_action( 'wp', array( $this, 'set_standard_keys' ), 100 );
 
 		// Set the default stale while revalidate and stale while error values.
```
Comment (simonwistow) [inline purgely.php]: > if error

File: src/settings-page.php [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+			'enable_stale_while_error',
+			__( 'Enable Stale While Error', 'purgely' ),
+			array( $this, 'enable_stale_while_error_render' ),
+			'purgely-settings',
+			'purgely-stale_settings'
+		);
+
+		add_settings_field(
+			'stale_while_error_ttl',
+			__( 'Stale While Error TTL (in Seconds)', 'purgely' ),
```
Comment (simonwistow) [inline src/settings-page.php]: > If Error

File: src/settings-page.php [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+	 */
+	public function enable_stale_while_error_render() {
+		$options = Purgely_Settings::get_settings();
+		?>
+		<input type='radio' name='purgely-settings[enable_stale_while_error]' <?php checked( isset( $options['enable_stale_while_error'] ) && true === $options['enable_stale_while_error'] ); ?> value='true'>Yes&nbsp;
+		<input type='radio' name='purgely-settings[enable_stale_while_error]' <?php checked( isset( $options['enable_stale_while_error'] ) && false === $options['enable_stale_while_error'] ); ?> value='false'>No
+		<p class="description">
+			<?php
+			printf(
+				esc_html__( 'Turn the "stale while error" behavior on or off. The stale while error behavior allows stale content to be served while the origin is returning an error state. Please see %s', 'purgely' ),
```
Comment (simonwistow) [inline src/settings-page.php]: > if error

File: src/settings-page.php [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+				sprintf(
+					'<a href="%1$s" target="_blank">%2$s</a>',
+					'https://www.fastly.com/blog/stale-while-revalidate',
+					esc_html__( 'Fastly\'s documentation for more information on stale while error', 'purgely' )
```
Comment (simonwistow) [inline src/settings-page.php]: > if error

File: purgely.php [hunk 9]
```diff
@@ -129,18 +134,18 @@ public function __construct() {
...
-		if ( true === PURGELY_ENABLE_STALE_WHILE_REVALIDATE ) {
-			$this->add_cache_control_header( PURGELY_STALE_WHILE_REVALIDATE_TTL, 'stale-while-revalidate' );
+		if ( true === Purgely_Settings::get_setting( 'enable_stale_while_revalidate' ) ) {
+			$this->add_cache_control_header( Purgely_Settings::get_setting( 'stale_while_revalidate_ttl' ), 'stale-while-revalidate' );
 		}
 
-		if ( true === PURGELY_ENABLE_STALE_WHILE_ERROR ) {
-			$this->add_cache_control_header( PURGELY_STALE_WHILE_ERROR_TTL, 'stale-while-error' );
+		if ( true === Purgely_Settings::get_setting( 'enable_stale_while_error' ) ) {
+			$this->add_cache_control_header( Purgely_Settings::get_setting( 'stale_while_error_ttl' ), 'stale-while-error' );
```
Comment (simonwistow) [inline purgely.php]: > if_error

**Comment by tollmanz** [comment]
Thanks @simonwistow! I've updated all refs to "stale while error". Best to get this right right now.

Note that this change will revert some users back to the default stale if error behavior, which is turned on and set for 24 hours. I don't imagine that will be too much of a problem.

**Pull request merged**